### DISCUSSION TITLE
Load asynchronous traits

### DISCUSSION THREAD

**Discussion by maalej** [GrapesJS/grapesjs#4933] [Q&A]
Hello,
I need to set traits asynchronously. On component initialization, I fetch traits from a server, however I don't find a way to add them to the component with addTrait.
Here is an example:
```
model: {
		    defaults: {
			    script(){
			      fetch(`http://localhost:4001/traits`)
		                .then((response) => response.json())
		                .then((data) => {
                                    editor.Components.getComponent('comp-image').addTrait(data[0]) 
		                })
			    }
		}
```
it says editor is not defined. Also I am unable to access directly the component traits with this.addTraits(data[0}).

Any help please?

**Comment by artf** [comment]
Hi @maalej as described [here](https://grapesjs.com/docs/modules/Components-js.html#important-caveat), everything inside the component `script` is outside of the editor context (the `script` is basically used to add JS to your final compiled project, like the [countdown component](https://github.com/GrapesJS/components-countdown), so it doesn't make sense to refer to the editor context).

What you probably need is to define that logic inside the init method
```js
model: {
    defaults: {...},

    init() {
      fetch(`http://localhost:4001/traits`)
        .then((response) => response.json())
        .then((data) => this.addTrait(data[0]))
    }
},
```
